File: apps/ios/GalaxySSI/AgentConfirmationPolicy.swift
import Foundation

enum AgentConfirmationTier: String, Codable, CaseIterable, Identifiable {
  case direct = "DIRECT"
  case confirmOnce = "CONFIRM_ONCE"
  case confirmAlways = "CONFIRM_ALWAYS"

  var id: String { rawValue }
}

enum AgentConfirmationPolicy {
  static func tier(for action: AgentAction) -> AgentConfirmationTier {
    let value = searchableValue(action)
    let toolId = nativeToolId(action)
    if toolId == homeAssistantServiceCall && requiresAlwaysHomeAssistantConfirmation(action.parameters["input_json"] ?? "") {
      return .confirmAlways
    }
    if alwaysConfirmNativeToolIds.contains(toolId) {
      return .confirmAlways
    }
    if confirmOnceNativeToolIds.contains(toolId) {
      return .confirmOnce
    }
    if desktopRemoteNativeToolIds.contains(toolId) {
      return .direct
    }
    if toolId == webSearch || webIntelligenceToolIds.contains(toolId) {
      return .direct
    }
    if alwaysConfirmKinds.contains(action.kind) || alwaysConfirmTerms.contains(where: value.contains) {
      return .confirmAlways
    }
    if action.kind == .callConnector {
      return .direct
    }
    if confirmOnceTerms.contains(where: value.contains) || action.kind == .controlDevice {
      return .confirmOnce
    }
    if action.kind == .setAlarm ||
      action.kind == .openApp ||
      directActionIds.contains(action.id) ||
      directNativeToolIds.contains(toolId) ||
      directTerms.contains(where: value.contains) {
      return .direct
    }
    switch action.risk {
    case .low:
      return .direct
    case .medium:
      return .confirmOnce
    case .high, .blocked:
      return .confirmAlways
    }
  }

  static func consentKey(for action: AgentAction) -> String {
    let value = searchableValue(action)
    let toolId = nativeToolId(action)
    if locationTerms.contains(where: value.contains) {
      return "location"
    }
    if microphoneTerms.contains(where: value.contains) {
      return "microphone"
    }
    if downloadTerms.contains(where: value.contains) {
      return "downloads"
    }
    if contactWriteTerms.contains(where: value.contains) {
      return "contacts_write"
    }
    if calendarWriteTerms.contains(where: value.contains) {
      return "calendar_write"
    }
    if toolId == bluetoothDiscoveryForeground {
      return "bluetooth_discovery"
    }
    if toolId == wifiScanStart {
      return "wifi_scan"
    }
    if toolId == installedAppsList || toolId == packageDetail {
      return "installed_apps_read"
    }
    if toolId == homeAssistantEntitiesList || toolId == homeAssistantEntityRead {
      return "home_assistant_read"
    }
    if toolId == homeAssistantServiceCall {
      return homeAssistantConsentScope(action.parameters["input_json"] ?? "")
    }
    if action.kind == .controlDevice {
      return "device_control:\(action.target.lowercased().trimmingCharacters(in: .whitespacesAndNewlines))"
    }
    return "action:\(action.kind.rawValue.lowercased()):\(action.id.lowercased().trimmingCharacters(in: .whitespacesAndNewlines))"
  }

  private static func nativeToolId(_ action: AgentAction) -> String {
    action.parameters["tool_id"] ?? ""
  }

  private static func searchableValue(_ action: AgentAction) -> String {
    var parts = [action.id, action.kind.rawValue, action.target, action.description]
    for (key, value) in action.parameters where !key.hasPrefix(internalParameterPrefix) {
      parts.append(key)
      parts.append(value)
    }
    return parts.joined(separator: " ").lowercased()
  }

  private static func requiresAlwaysHomeAssistantConfirmation(_ inputJson: String) -> Bool {
    let input = homeAssistantInput(inputJson)
    let cleanEntity = input.entityId.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    let entityDomain = cleanEntity.split(separator: ".", maxSplits: 1).first.map(String.init) ?? ""
    let identity = "\(cleanEntity) \(input.serviceDomain.lowercased()) \(input.service.lowercased())"
    return homeAssistantAlwaysConfirmDomains.contains(entityDomain) ||
      homeAssistantAlwaysConfirmServices.contains(input.service.lowercased()) ||
      homeAssistantAlwaysConfirmIdentityTerms.contains(where: identity.contains)
  }

  private static func homeAssistantConsentScope(_ inputJson: String) -> String {
    let entityId = homeAssistantInput(inputJson).entityId
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .lowercased()
    if entityId.range(of: #"^[a-z0-9_]+\.[a-z0-9_]+$"#, options: .regularExpression) != nil {
      return "home_assistant_control:\(entityId)"
    }
    return "home_assistant_control"
  }

  private static func homeAssistantInput(_ inputJson: String) -> (entityId: String, serviceDomain: String, service: String) {
    guard let data = inputJson.data(using: .utf8),
          let rawObject = try? JSONSerialization.jsonObject(with: data),
          let object = rawObject as? [String: Any] else {
      return ("", "", "")
    }
    return (
      object["entity_id"] as? String ?? "",
      object["service_domain"] as? String ?? "",
      object["service"] as? String ?? ""
    )
  }

  private static let internalParameterPrefix = "_galaxyssi_"
  private static let homeAssistantServiceCall = "galaxyssi.home_assistant.service.call"
  private static let homeAssistantEntitiesList = "galaxyssi.home_assistant.entities.list"
  private static let homeAssistantEntityRead = "galaxyssi.home_assistant.entity.read"
  private static let bluetoothDiscoveryForeground = "galaxyssi.hardware.bluetooth.discovery.foreground"
  private static let installedAppsList = "galaxyssi.hardware.apps.installed.list"
  private static let packageDetail = "galaxyssi.hardware.apps.package.detail"
  private static let wifiScanStart = "galaxyssi.android.wifi.scan.start"
  private static let webSearch = "web.search"

  private static let alwaysConfirmKinds: Set<AgentActionKind> = [.replyNotification, .deleteText, .lockScreen]
  private static let directActionIds: Set<String> = [
    "set-timer", "open-timer", "set-alarm", "open-camera", "open-flashlight",
    "battery-status", "device-status"
  ]
  private static let desktopRemoteNativeToolIds: Set<String> = [
    "galaxyssi.desktop.windows.system.status",
    "galaxyssi.desktop.windows.process.list",
    "galaxyssi.desktop.workspace.file.list",
    "galaxyssi.desktop.workspace.file.read.text",
    "galaxyssi.desktop.workspace.file.write.text",
    "galaxyssi.desktop.workspace.file.sha256",
    "galaxyssi.desktop.workspace.archive.create",
    "galaxyssi.desktop.terminal.run",
    "galaxyssi.desktop.office.document.inspect",
    "galaxyssi.desktop.office.document.convert"
  ]
  private static let webIntelligenceToolIds: Set<String> = [
    "galaxyssi.web.intelligence.search",
    "galaxyssi.web.intelligence.fetch",
    "galaxyssi.web.intelligence.crawl",
    "galaxyssi.web.intelligence.extract",
    "galaxyssi.web.intelligence.cache",
    "galaxyssi.web.intelligence.find_similar",
    "galaxyssi.web.intelligence.research",
    "galaxyssi.web.intelligence.agent",
    "galaxyssi.web.intelligence.diff",
    "galaxyssi.web.intelligence.watch"
  ]
  private static let directNativeToolIds = Set([
    "galaxyssi.hardware.device.status",
    "galaxyssi.hardware.battery.status",
    "galaxyssi.hardware.power.status",
    "galaxyssi.hardware.memory.status",
    "galaxyssi.hardware.storage.status",
    "galaxyssi.hardware.network.status",
    "galaxyssi.hardware.sensors.list",
    "galaxyssi.hardware.sensor.sample",
    "galaxyssi.hardware.bluetooth.status",
    "galaxyssi.hardware.nfc.status",
    "galaxyssi.hardware.flashlight.set",
    "galaxyssi.camera.capture.visible",
    "web.search",
    "galaxyssi.media.ffmpeg.transcode",
    "galaxyssi.runtime.execute",
    "galaxyssi.hardware.bluetooth.pairing.handoff",
    "galaxyssi.android.audio.status",
    "galaxyssi.android.audio.volume.set",
    "galaxyssi.android.audio.mute.set",
    "galaxyssi.android.wifi.panel.open",
    "galaxyssi.android.wifi.hotspot.panel.open",
    "galaxyssi.android.biometric.enrollment.open"
  ])
    .union(webIntelligenceToolIds)
    .union(AgentIOSWebMediaNativeToolCatalog.directToolIds)
  private static let confirmOnceNativeToolIds: Set<String> = Set([
    "galaxyssi.microphone.record.visible",
    "galaxyssi.notifications.list",
    bluetoothDiscoveryForeground,
    installedAppsList,
    packageDetail,
    wifiScanStart,
    "galaxyssi.runtime.packs.install",
    homeAssistantEntitiesList,
    homeAssistantEntityRead,
    homeAssistantServiceCall
  ]).union(AgentIOSWebMediaNativeToolCatalog.confirmOnceToolIds)
  private static let alwaysConfirmNativeToolIds: Set<String> = [
    "galaxyssi.notifications.reply",
    "galaxyssi.desktop.terminal.run"
  ]

  private static let alwaysConfirmTerms = [
    "send sms", "sms.send", "reply sms", "send message", "reply message", "reply notification",
    "send email", "reply email", "phone call", "dial", "telephony.dial", "delete", "remove",
    "install", "uninstall", "payment", "purchase", "checkout", "transfer", "grant permission",
    "authorize", "security setting", "system setting", "registry edit", "screen lock", "lock device",
    "device_policy.lock", "reboot", "git commit", "git push", "submit code", "commit code",
    "door lock", "smart lock", "garage door", "alarm panel", "private key", "password",
    "\u{53D1}\u{9001}\u{77ED}\u{4FE1}", "\u{56DE}\u{590D}\u{77ED}\u{4FE1}",
    "\u{53D1}\u{6D88}\u{606F}", "\u{56DE}\u{590D}\u{6D88}\u{606F}",
    "\u{6253}\u{7535}\u{8BDD}", "\u{62E8}\u{53F7}", "\u{5220}\u{9664}",
    "\u{5B89}\u{88C5}", "\u{5378}\u{8F7D}", "\u{652F}\u{4ED8}",
    "\u{8F6C}\u{8D26}", "\u{6388}\u{6743}", "\u{6743}\u{9650}",
    "\u{5B89}\u{5168}\u{8BBE}\u{7F6E}", "\u{7CFB}\u{7EDF}\u{8BBE}\u{7F6E}",
    "\u{63D0}\u{4EA4}\u{4EE3}\u{7801}", "\u{63A8}\u{9001}\u{4EE3}\u{7801}",
    "\u{9501}\u{5C4F}", "\u{91CD}\u{542F}", "\u{95E8}\u{9501}", "\u{8F66}\u{5E93}\u{95E8}"
  ]
  private static let directTerms = [
    "timer", "alarm clock", "set alarm", "camera capture", "take photo", "flashlight", "torch",
    "audio volume", "set volume", "audio mute", "open app", "launch app", "battery status",
    "device status", "read battery", "read device", "\u{8BA1}\u{65F6}\u{5668}",
    "\u{95F9}\u{949F}", "\u{62CD}\u{7167}", "\u{624B}\u{7535}\u{7B52}",
    "\u{97F3}\u{91CF}", "\u{6253}\u{5F00}app", "\u{6253}\u{5F00} app",
    "\u{7535}\u{91CF}", "\u{8BBE}\u{5907}\u{72B6}\u{6001}"
  ]
  private static let locationTerms = ["location", "gps", "\u{5B9A}\u{4F4D}", "\u{4F4D}\u{7F6E}"]
  private static let microphoneTerms = ["microphone", "record audio", "\u{9EA6}\u{514B}\u{98CE}", "\u{5F55}\u{97F3}"]
  private static let downloadTerms = ["download", "\u{4E0B}\u{8F7D}"]
  private static let contactWriteTerms = [
    "contacts.write", "contact upsert", "create contact", "update contact",
    "\u{65B0}\u{5EFA}\u{8054}\u{7CFB}\u{4EBA}", "\u{4FEE}\u{6539}\u{8054}\u{7CFB}\u{4EBA}",
    "\u{66F4}\u{65B0}\u{8054}\u{7CFB}\u{4EBA}"
  ]
  private static let calendarWriteTerms = [
    "calendar.write", "calendar event upsert", "create calendar event", "update calendar event",
    "\u{65B0}\u{5EFA}\u{65E5}\u{7A0B}", "\u{4FEE}\u{6539}\u{65E5}\u{7A0B}",
    "\u{66F4}\u{65B0}\u{65E5}\u{7A0B}"
  ]
  private static let confirmOnceTerms = locationTerms + microphoneTerms + downloadTerms + contactWriteTerms + calendarWriteTerms
  private static let homeAssistantAlwaysConfirmDomains: Set<String> = [
    "alarm_control_panel", "automation", "camera", "lock", "script", "siren", "valve"
  ]
  private static let homeAssistantAlwaysConfirmServices: Set<String> = [
    "alarm_arm_away", "alarm_arm_home", "alarm_arm_night", "alarm_disarm", "alarm_trigger", "unlock"
  ]
  private static let homeAssistantAlwaysConfirmIdentityTerms = [
    "alarm", "door", "gate", "garage", "lock", "security", "siren"
  ]
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentConfirmationPolicy.kt
package com.galaxyssi.chat

import android.content.Context

enum class AgentConfirmationTier {
    DIRECT,
    CONFIRM_ONCE,
    CONFIRM_ALWAYS
}

enum class AgentPermissionChoice(val wireValue: String) {
    ALLOW_ONCE("allow_once"),
    ALLOW_SESSION("allow_session"),
    ALLOW_ALWAYS("allow_always"),
    DENY_ALWAYS("deny_always");

    val approved: Boolean
        get() = this != DENY_ALWAYS

    companion object {
        fun fromWireValue(value: String): AgentPermissionChoice? =
            entries.firstOrNull { it.wireValue == value.trim().lowercase() }
    }
}

data class AgentConfirmationConsentDecision(
    val allowed: Boolean,
    val denied: Boolean,
    val choice: AgentPermissionChoice? = null,
    val grantId: String = "",
    val reason: String = ""
)

interface AgentConfirmationConsentStore {
    fun decision(
        consentKey: String,
        sessionId: String = "",
        consume: Boolean = false
    ): AgentConfirmationConsentDecision
    fun rememberedKeys(): Set<String>
    fun record(
        consentKey: String,
        choice: AgentPermissionChoice,
        sessionId: String = ""
    ): AgentConfirmationConsentDecision
    fun forget(consentKey: String): Boolean
    fun endSession(sessionId: String)
    fun clear()
}

class SharedPreferencesAgentConfirmationConsentStore(context: Context) : AgentConfirmationConsentStore {
    private val appContext = context.applicationContext
    private val grantStore: AgentPermissionGrantStore = EncryptedAgentPermissionGrantStore(appContext)
    private val revocationCoordinator by lazy {
        AgentPermissionRevocationCoordinator(
            grantStore = grantStore,
            workspaceStore = EncryptedAgentWorkspaceStore(appContext),
            runEventStore = AgentRunEventStore(appContext),
            pauseActiveWorkspace = { workspaceId, reason ->
                AgentTaskRuntime.supervisor(appContext).pauseForPermissionRevocation(workspaceId, reason)
            }
        )
    }

    override fun decision(
        consentKey: String,
        sessionId: String,
        consume: Boolean
    ): AgentConfirmationConsentDecision {
        val cleanKey = consentKey.trim()
        if (cleanKey.isBlank()) return AgentConfirmationConsentDecision(
            allowed = false,
            denied = false,
            reason = "invalid_consent_key"
        )
        val result = grantStore.authorize(
            permissionRequest(cleanKey, sessionId),
            consume = consume
        )
        val grant = result.grant
        return AgentConfirmationConsentDecision(
            allowed = result.granted,
            denied = grant?.effect == AgentPermissionGrantEffect.DENY,
            choice = grant?.toPermissionChoice(),
            grantId = grant?.grantId.orEmpty(),
            reason = result.reason
        )
    }

    override fun rememberedKeys(): Set<String> = grantStore.list(includeInactive = false)
        .asSequence()
        .filter { it.subjectType == AgentPermissionSubjectType.CONSEQUENTIAL_ACTION }
        .filter { it.effect == AgentPermissionGrantEffect.ALLOW }
        .map(AgentPermissionGrant::scope)
        .toSet()

    override fun record(
        consentKey: String,
        choice: AgentPermissionChoice,
        sessionId: String
    ): AgentConfirmationConsentDecision {
        val cleanKey = consentKey.trim()
        if (cleanKey.isBlank()) return AgentConfirmationConsentDecision(
            allowed = false,
            denied = false,
            reason = "invalid_consent_key"
        )
        val cleanSessionId = sessionId.trim()
        if (choice == AgentPermissionChoice.ALLOW_SESSION) {
            require(cleanSessionId.isNotBlank()) { "Session permission requires a session id" }
        }
        val before = rememberedKeys()
        grantStore.revokeScope(cleanKey, "user_replaced_permission_decision")
        val lifetime = when (choice) {
            AgentPermissionChoice.ALLOW_ONCE -> AgentPermissionGrantLifetime.SINGLE_USE
            AgentPermissionChoice.ALLOW_SESSION -> AgentPermissionGrantLifetime.SESSION
            AgentPermissionChoice.ALLOW_ALWAYS,
            AgentPermissionChoice.DENY_ALWAYS -> AgentPermissionGrantLifetime.PERMANENT
        }
        val effect = if (choice.approved) {
            AgentPermissionGrantEffect.ALLOW
        } else {
            AgentPermissionGrantEffect.DENY
        }
        val grant = grantStore.grant(
            AgentPermissionGrant(
                subjectType = AgentPermissionSubjectType.CONSEQUENTIAL_ACTION,
                subjectId = HOST_SUBJECT_ID,
                scope = cleanKey,
                action = cleanKey,
                issuer = AgentPermissionGrantIssuer.USER,
                evidence = "user_decision:${choice.wireValue}",
                lifetime = lifetime,
                effect = effect,
                sessionId = if (lifetime == AgentPermissionGrantLifetime.SESSION) cleanSessionId else ""
            )
        )
        val current = if (choice == AgentPermissionChoice.ALLOW_ONCE) {
            grantStore.authorize(permissionRequest(cleanKey, cleanSessionId), consume = true)
        } else {
            AgentPermissionDecision(choice.approved, grant, "user_decision_recorded")
        }
        val after = rememberedKeys()
        GlobalConversationEventBus.publishCapabilityEvents(
            appContext,
            GlobalCapabilityObservationExtractor.authorizationMutations(before, after)
        )
        return AgentConfirmationConsentDecision(
            allowed = current.granted,
            denied = effect == AgentPermissionGrantEffect.DENY,
            choice = choice,
            grantId = grant.grantId,
            reason = current.reason
        )
    }

    override fun forget(consentKey: String): Boolean {
        val cleanKey = consentKey.trim()
        val before = rememberedKeys()
        val revocation = revocationCoordinator.revokeScope(
            cleanKey,
            "user_revoked_remembered_confirmation"
        ).revocation
        if (revocation.revokedGrantIds.isEmpty()) return false
        val after = rememberedKeys()
        GlobalConversationEventBus.publishCapabilityEvents(
            appContext,
            GlobalCapabilityObservationExtractor.authorizationMutations(before, after)
        )
        return true
    }

    override fun endSession(sessionId: String) {
        val cleanSessionId = sessionId.trim()
        if (cleanSessionId.isBlank()) return
        val before = rememberedKeys()
        grantStore.revokeSession(cleanSessionId, "agent_session_ended")
        val after = rememberedKeys()
        GlobalConversationEventBus.publishCapabilityEvents(
            appContext,
            GlobalCapabilityObservationExtractor.authorizationMutations(before, after)
        )
    }

    override fun clear() = grantStore.clear()

    private fun permissionRequest(consentKey: String, sessionId: String) = AgentPermissionRequest(
        subjectType = AgentPermissionSubjectType.CONSEQUENTIAL_ACTION,
        subjectId = HOST_SUBJECT_ID,
        scope = consentKey,
        action = consentKey,
        sessionId = sessionId
    )

    private fun AgentPermissionGrant.toPermissionChoice(): AgentPermissionChoice = when {
        effect == AgentPermissionGrantEffect.DENY -> AgentPermissionChoice.DENY_ALWAYS
        lifetime == AgentPermissionGrantLifetime.SINGLE_USE -> AgentPermissionChoice.ALLOW_ONCE
        lifetime == AgentPermissionGrantLifetime.SESSION -> AgentPermissionChoice.ALLOW_SESSION
        else -> AgentPermissionChoice.ALLOW_ALWAYS
    }

    private companion object {
        const val HOST_SUBJECT_ID = "galaxyssi-host"
    }
}

object AgentConfirmationPolicy {
    fun tier(@Suppress("UNUSED_PARAMETER") action: AgentAction): AgentConfirmationTier =
        AgentConfirmationTier.DIRECT

    fun consentKey(action: AgentAction): String {
        val value = searchableValue(action)
        return when {
            LOCATION_TERMS.any(value::contains) -> "location"
            MICROPHONE_TERMS.any(value::contains) -> "microphone"
            DOWNLOAD_TERMS.any(value::contains) -> "downloads"
            CONTACT_WRITE_TERMS.any(value::contains) -> "contacts_write"
            CALENDAR_WRITE_TERMS.any(value::contains) -> "calendar_write"
            nativeToolId(action) == AgentHardwareNativeTools.BLUETOOTH_DISCOVERY_FOREGROUND -> "bluetooth_discovery"
            nativeToolId(action) == AgentAndroidSystemNativeTools.WIFI_SCAN_START -> "wifi_scan"
            nativeToolId(action) in setOf(
                AgentHardwareNativeTools.INSTALLED_APPS_LIST,
                AgentHardwareNativeTools.PACKAGE_DETAIL
            ) -> "installed_apps_read"
            nativeToolId(action) in setOf(
                AgentHomeAssistantNativeTools.ENTITIES_LIST,
                AgentHomeAssistantNativeTools.ENTITY_READ
            ) -> "home_assistant_read"
            nativeToolId(action) == AgentHomeAssistantNativeTools.SERVICE_CALL ->
                AgentHomeAssistantNativeTools.consentScope(action.parameters["input_json"].orEmpty())
            action.kind == AgentActionKind.CONTROL_DEVICE -> "device_control:${action.target.lowercase().trim()}"
            else -> "action:${action.kind.name.lowercase()}:${action.id.lowercase().trim()}"
        }
    }

    private fun nativeToolId(action: AgentAction): String = action.parameters["tool_id"].orEmpty()

    private fun searchableValue(action: AgentAction): String = buildString {
        append(action.id).append(' ')
        append(action.kind.name).append(' ')
        append(action.target).append(' ')
        append(action.description).append(' ')
        action.parameters.forEach { (key, value) ->
            // Session context is evidence for the model, not part of the current action's risk intent.
            if (!key.startsWith(INTERNAL_PARAMETER_PREFIX)) {
                append(key).append(' ').append(value).append(' ')
            }
        }
    }.lowercase()

    private const val INTERNAL_PARAMETER_PREFIX = "_galaxyssi_"

    private val ALWAYS_CONFIRM_KINDS = setOf(
        AgentActionKind.REPLY_NOTIFICATION,
        AgentActionKind.DELETE_TEXT,
        AgentActionKind.LOCK_SCREEN
    )

    private val DIRECT_ACTION_IDS = setOf(
        "set-timer", "open-timer", "set-alarm", "open-camera", "open-flashlight",
        "battery-status", "device-status"
    )

    private val DIRECT_NATIVE_TOOL_IDS = setOf(
        AgentHardwareNativeTools.BATTERY_STATUS,
        AgentHardwareNativeTools.POWER_STATUS,
        AgentHardwareNativeTools.MEMORY_STATUS,
        AgentHardwareNativeTools.STORAGE_STATUS,
        AgentHardwareNativeTools.NETWORK_STATUS,
        AgentHardwareNativeTools.SENSORS_LIST,
        AgentHardwareNativeTools.SENSOR_SAMPLE,
        AgentHardwareNativeTools.BLUETOOTH_STATUS,
        AgentHardwareNativeTools.NFC_STATUS,
        AgentHardwareNativeTools.FLASHLIGHT_SET,
        AgentVisibleCaptureNativeTools.CAMERA_CAPTURE,
        AgentWebMediaNativeTools.WEB_SEARCH,
        *AgentWebIntelligenceNativeTools.toolIds.toTypedArray(),
        AgentWebMediaNativeTools.MEDIA_FFMPEG_TRANSCODE,
        AgentOnDeviceRuntimeTools.EXECUTE,
        AgentHardwareNativeTools.BLUETOOTH_PAIRING_HANDOFF,
        AgentAndroidSystemNativeTools.AUDIO_STATUS,
        AgentAndroidSystemNativeTools.AUDIO_VOLUME_SET,
        AgentAndroidSystemNativeTools.AUDIO_MUTE_SET,
        AgentAndroidSystemNativeTools.WIFI_PANEL_OPEN,
        AgentAndroidSystemNativeTools.WIFI_HOTSPOT_PANEL_OPEN,
        AgentAndroidSystemNativeTools.BIOMETRIC_ENROLLMENT_OPEN
    )

    private val CONFIRM_ONCE_NATIVE_TOOL_IDS = setOf(
        AgentVisibleCaptureNativeTools.MICROPHONE_RECORD,
        AgentNotificationNativeTools.NOTIFICATIONS_LIST,
        AgentHardwareNativeTools.BLUETOOTH_DISCOVERY_FOREGROUND,
        AgentHardwareNativeTools.INSTALLED_APPS_LIST,
        AgentHardwareNativeTools.PACKAGE_DETAIL,
        AgentAndroidSystemNativeTools.WIFI_SCAN_START,
        AgentOnDeviceRuntimeTools.INSTALL_PACK,
        AgentHomeAssistantNativeTools.ENTITIES_LIST,
        AgentHomeAssistantNativeTools.ENTITY_READ,
        AgentHomeAssistantNativeTools.SERVICE_CALL
    )

    private val ALWAYS_CONFIRM_NATIVE_TOOL_IDS = setOf(
        AgentNotificationNativeTools.NOTIFICATION_REPLY
    )

    private val ALWAYS_CONFIRM_TERMS = listOf(
        "send sms", "sms.send", "reply sms", "send message", "reply message", "reply notification",
        "send email", "reply email", "phone call", "dial", "telephony.dial", "delete", "remove",
        "install", "uninstall", "payment", "purchase", "checkout", "transfer", "grant permission",
        "authorize", "security setting", "system setting", "registry edit", "screen lock", "lock device",
        "device_policy.lock", "reboot", "git commit", "git push", "submit code", "commit code",
        "door lock", "smart lock", "garage door", "alarm panel", "private key", "password",
        "\u53d1\u9001\u77ed\u4fe1", "\u56de\u590d\u77ed\u4fe1", "\u53d1\u6d88\u606f", "\u56de\u590d\u6d88\u606f",
        "\u6253\u7535\u8bdd", "\u62e8\u53f7", "\u5220\u9664", "\u5b89\u88c5", "\u5378\u8f7d", "\u652f\u4ed8",
        "\u8f6c\u8d26", "\u6388\u6743", "\u6743\u9650", "\u5b89\u5168\u8bbe\u7f6e", "\u7cfb\u7edf\u8bbe\u7f6e",
        "\u63d0\u4ea4\u4ee3\u7801", "\u63a8\u9001\u4ee3\u7801", "\u9501\u5c4f", "\u91cd\u542f",
        "\u95e8\u9501", "\u8f66\u5e93\u95e8"
    )

    private val DIRECT_TERMS = listOf(
        "timer", "alarm clock", "set alarm", "camera capture", "take photo", "flashlight", "torch",
        "audio volume", "set volume", "audio mute", "open app", "launch app", "battery status",
        "device status", "read battery", "read device", "\u8ba1\u65f6\u5668", "\u95f9\u949f", "\u62cd\u7167",
        "\u624b\u7535\u7b52", "\u97f3\u91cf", "\u6253\u5f00app", "\u6253\u5f00 app", "\u7535\u91cf", "\u8bbe\u5907\u72b6\u6001"
    )

    private val LOCATION_TERMS = listOf("location", "gps", "\u5b9a\u4f4d", "\u4f4d\u7f6e")
    private val MICROPHONE_TERMS = listOf("microphone", "record audio", "\u9ea6\u514b\u98ce", "\u5f55\u97f3")
    private val DOWNLOAD_TERMS = listOf("download", "\u4e0b\u8f7d")
    private val CONTACT_WRITE_TERMS = listOf(
        "contacts.write", "contact upsert", "create contact", "update contact",
        "\u65b0\u5efa\u8054\u7cfb\u4eba", "\u4fee\u6539\u8054\u7cfb\u4eba", "\u66f4\u65b0\u8054\u7cfb\u4eba"
    )
    private val CALENDAR_WRITE_TERMS = listOf(
        "calendar.write", "calendar event upsert", "create calendar event", "update calendar event",
        "\u65b0\u5efa\u65e5\u7a0b", "\u4fee\u6539\u65e5\u7a0b", "\u66f4\u65b0\u65e5\u7a0b"
    )
    private val CONFIRM_ONCE_TERMS =
        LOCATION_TERMS + MICROPHONE_TERMS + DOWNLOAD_TERMS + CONTACT_WRITE_TERMS + CALENDAR_WRITE_TERMS
}
